core/blockchain.go

```diff
@@ -1340,7 +1340,7 @@ func (bc *BlockChain) insertSidechain(it *insertIterator) (int, []interface{}, [
 		numbers []uint64
 	)
 	parent := bc.GetHeader(it.previous().Hash(), it.previous().NumberU64())
-	for parent != nil && !bc.HasState(parent.Root) {
+	for parent != nil && rawdb.ReadCanonicalHash(bc.db, parent.Number.Uint64()) != parent.Hash() {
```

Comment:
The intent of this loop is not to find the common ancestor, but to find the first block in our history where we have the actual state. So that may be genesis, or some other point where we flushed the state to disk. This change turns the loop into something completely different